configs/config.prod.yml

```diff
@@ -24,6 +25,27 @@ experiment:
   # split corpus to parallelize translation
   split-length: 2000000
 
+  best-model: chrf
+
+  bicleaner:
+    default-threshold: 0.5
+    dataset-thresholds:
+      opus_CCAligned/v1: 0.7
+      opus_WikiMatrix/v1: 0.7
+      opus_OpenSubtitles/v2018: 0.9
+      opus_bible-uedin/v1: 0.7
+      mtdata_cc_aligned: 0.7
+      mtdata_wiki_titles_v1: 0.7
+      mtdata_WikiMatrix_v1: 0.7
+      mtdata_wiki_titles_v2: 0.7
+      mtdata_wmt13_commoncrawl: 0.7
+
+training:
+  s2s:
+    after-epochs: 10
+  # limit number of epochs on augmented dataset (high resource languages)
+  teacher-all:
+    after-epochs: 5
```

Comment:
This is not adequate for lower resource languages, that need quite a few more epochs to converge.

Reply:
Yes, that's why I moved it to the main config, to be able to adjust based on low resource/high resource. For low resource, I would just remove this and train until early stopping. I can add comments about this.

Does fine-tuning on very limited parallel data even make sense for low resource?

Is 5 epochs a good number for high resource languages?

Reply:
I would say that there isn't a hard and fast rule about the number of epochs. It really depends on how varied your data is. Eg if your data is from the same domain and very similar, maybe it will be done in 5 epochs. If the data is diverse, then maybe not. I prefer for humans to determine the patience. Is a researcher able to login and `ctrl+c` this step and continue to the next step?

Reply:
The logic here is that we preconfigure a number of epochs for training on a dataset augmented with back translations then continue training on parallel data only with early stopping (advice from Kenneth: "train on backtranslated first for several epochs, then continue on parallel".).  If we want to pretrain on augmented dataset to convergence, it also should use early stopping.

I don't think it's easy to skip things with snakemake. If a step fails it deletes the expected output to prevent corrupted data. It is possible to stop the workflow, fix some things and continue from where it stopped. (harder on Slurm because it requires cancelling Slurm jobs manually). It is also possible to run snakemake with a specific target to execute only a part of the pipeline.

Reply:
The pretraining on backtranslation is controlled by epochs and the continued training on parallel is controlled by patience then?

Reply:
Yes and this is controlled in the config, so if we remove `after-epochs: 5`, it will be also trained until early stopped (10 epochs by default). I assumed a user can remove it for low resource but keep fixed for high resource language. I can remove it if you think it makes more sense to train on augmented data until convergence.

Reply:
As Nick points out, reasonable settings of epochs depend on 1. the total amount of training data available, and 2. the diversity of the training data. May I suggest that instead of `after-epochs` (which is a deprecated parameter anyway; the preferred way of specifying a stopping point is `after`, which allows us to specify the number of updates or the amount of training data in terms of target labels rather than the number of iterations over the available data. See https://marian-nmt.github.io/docs/cmd/marian/). And no, I don't know what a reasonable value is. This would require experimentation (if such experimentation has been done already, I'm not aware of it, but we have plenty of training logs where we could look at when the learning curve starts to flatten out.

Reply:
I moved these settings to `pipeline/train/configs` and updated them to use `after`. It will be easier to modify everything in one place.